base-files: add the possibility to set sysctl via the uci

feckert opened the pull request:
Sometimes it is necessary that certain settings in the sysctl can be easily changed via uci. In my case it is about the `/proc/sys/net/ipv4/fib_multipath_hash_policy`. I would like to make this change via the LuCI. With the current interface via `/etc/sysctl.d/`, this is not so easy possible. If the settings are mapped as uci, it is much easier via LuCI to change this sysctl settings.
The service has been changed to procd so that the uci changes are executed during a `/sbin/reload_config`. So that the default value can be set by the kernel again when the sysctl section is disabled, the sysctl settings are read out once during bootup and stored in the `tmp` directory.

Unamed uci section in `/etc/config/sysctl`:
https://www.kernel.org/doc/html/latest/networking/ip-sysctl.html
```
config sysctl
            option enable '1'
            option section 'net.ipv4.fib_multipath_hash_policy'
            option value '2'
```

feckert committed:
base-files: save default sysctl on start

In order to get the default value of the sysctl, a reboot must be done
beforehand or it must be saved in the linux source code what the
default value is.

Reboot is not an option and the linux does not provide the feature to
save the default values.

To simplify this, all values of '/proc/sys' are saved on sysctl start
with the 'sysctl' command into a temp file after the default values from
OpenWrt are applied and before any changes are made by files that are
locate in '/etc/sysctl.conf' and '/etc/sysctl.d/*'.

Signed-off-by: Florian Eckert <fe@dev.tdt.de>

feckert committed:
base-files: add the possibility to set sysctl via uci

With this change, sysctl parameters can be set via the uci. The
following unamed section musst be set in '/etc/config/sysctl'.

config sysctl
	option enable '1'
	option section 'net.ipv4.fib_multipath_hash_policy'
	option value '2'

If the unamed section gets disabled, then the default value gets set,
that was saved during boot.

Signed-off-by: Florian Eckert <fe@dev.tdt.de>

feckert committed:
base-files: use procd for sysctl

Switch init to use procd, and also add a reload trigger if the config
file '/etc/config/sysctl' gets changed.

Signed-off-by: Florian Eckert <fe@dev.tdt.de>

Ansuel commented:
@feckert are you still interested in this?

I would make 2 changes to this...
one is to put these in the system config. (we have there also stuff related to ntp and leds... )

Also I would drop the enable thing. If the entry doesn't need to be enabled, we just remove it.

Ansuel commented on package/base-files/files/etc/init.d/sysctl:
```diff
@@ -36,9 +37,50 @@ apply_defaults() {
+	rm -f /tmp/sysctl.d/01-uci.conf
+
+	config_load sysctl
+	config_foreach handle_sysctl sysctl "/tmp/sysctl.d/01-uci.conf"
```
Why do we need all this complexity with tmp? Can't we just apply the value directly after the sysctl ones are applied?

feckert replied:
If someone deletes the configuration, then I thought the default/startup value should be set again as before the change.

Otherwise I have to reboot every time I delete the value in the uci configuration so that I have the default value of Kernel again.

Ansuel replied:
ugh sad that they never implemented a way in sysctl to reset a value... I still thing the logic can be reduced to just the line you added

`[ -f /tmp/sysctl-default.conf ] || sysctl -a 1>/tmp/sysctl-default.conf 2>/dev/null`

and change it to something like

```
if [ -f /tmp/sysctl-default.conf ]; then
  sysctl -p /tmp/sysctl-default.conf
else
  sysctl -a 1>/tmp/sysctl-default.conf 2>/dev/null
fi
```

We store in sysctl-default.conf the default values and we just repply them before any modification from packages and uci. And with this change, we can just parse the uci option and apply them directly.

Also I think we should cache the default sysctl after apply_defaults.

Or even move the apply_defaults and the sysctl-default.conf creation in start_service. And make reload_service only apply the file.

feckert commented:
> @feckert are you still interested in this?

yes

> I would make 2 changes to this... one is to put these in the system config. (we have there also stuff related to ntp and leds... )

I'm happy to do that. I'll have to take another look.

> Also I would drop the enable thing. If the entry doesn't need to be enabled, we just remove it.

I'm happy to do that. I'll have to take another look.

feckert committed:
base-files: reset values on sysctl stop

If the sysctl should stop, then set alle values to default.

Signed-off-by: Florian Eckert <fe@dev.tdt.de>

feckert committed:
base-files: also read sysctl from /tmp/run/sysctl.d

There are situations where the change of a sysctl parameter should not
be saved persistently. Therefore, this change adds the possibility that
sysctl parameters can also be loaded from '/tmp/run/sysctl.d/*.conf'.

Signed-off-by: Florian Eckert <fe@dev.tdt.de>

feckert commented:
@Ansuel I have update my PR and made the following changes to my.

-  Change the handling of `/tmp/sysctl-default.conf` generation, so that the file is generated at boot time after the default changes have been applied with the function `apply-defaults`
- Read files sysctl files from `/tmp/run/sysctl.d/*.conf`
- Add a service stop command that reapply the `/tmp/sysctl-default.con`

You also requested the following changes that which have not/not yet been implemented as I still have comments on them.

> one is to put these in the system config. (we have there also stuff related to ntp and leds... )

The problem is that if you make a change that does not affect sysctl, such as changing the hostname in `/etc/config/system`, the sysctl uci option will still be reapplied! Which is not good from my point of view.

> Also I would drop the enable thing. If the entry doesn't need to be enabled, we just remove it.

But then the default value is no longer set as with the bootup, the configured value in the systctl still remains! Additional in other uci configuration files/section also have the disable uci feature. So If I do not want to apply this configuration I could set disable flag and do not have to remove the whole uci configuration section. I think this is a good useability.

Ansuel commented:
> @Ansuel I have update my PR and made the following changes to my.
> 
> -  Change the handling of `/tmp/sysctl-default.conf` generation, so that the file is generated at boot time after the default changes have been applied with the function `apply-defaults`
> - Read files sysctl files from `/tmp/run/sysctl.d/*.conf`
> - Add a service stop command that reapply the `/tmp/sysctl-default.conf`
> 
> You also requested the following changes that which have not/not yet been implemented as I still have comments on them.
> 
> > one is to put these in the system config. (we have there also stuff related to ntp and leds... )
> 
> The problem is that if you make a change that does not affect sysctl, such as changing the hostname in `/etc/config/system`, the sysctl uci option will still be reapplied! Which is not good from my point of view.
> 

Ah! I didn't think of that :( but is it really bad? Sysctl are not ignored if the same value is applied?

Still a little thing but I would love to not have lots of config. 

> > Also I would drop the enable thing. If the entry doesn't need to be enabled, we just remove it.
> 
> But then the default value is no longer set as with the bootup, the configured value in the systctl still remains! Additional in other uci configuration files/section also have the disable uci feature. So If I do not want to apply this configuration I could set disable flag and do not have to remove the whole uci configuration section. I think this is a good useability.
>

feckert commented:
> > @Ansuel I have update my PR and made the following changes to my.
> > 
> > * Change the handling of `/tmp/sysctl-default.conf` generation, so that the file is generated at boot time after the default changes have been applied with the function `apply-defaults`
> > * Read files sysctl files from `/tmp/run/sysctl.d/*.conf`
> > * Add a service stop command that reapply the `/tmp/sysctl-default.conf`
> > 
> > You also requested the following changes that which have not/not yet been implemented as I still have comments on them.
> > > one is to put these in the system config. (we have there also stuff related to ntp and leds... )
> > 
> > 
> > The problem is that if you make a change that does not affect sysctl, such as changing the hostname in `/etc/config/system`, the sysctl uci option will still be reapplied! Which is not good from my point of view.
> 
> Ah! I didn't think of that :( but is it really bad? Sysctl are not ignored if the same value is applied?
> 
> Still a little thing but I would love to not have lots of config.

I have now looked at this in the [kernel](https://elixir.bootlin.com/linux/latest/source/kernel/sysctl.c). If I understand this correctly, it does not check whether the option is different. If the value that is configured is the same, then the same value is written again.

It's up to you what you want. Currently, if I change the hostname, the LEDs are reconfigured because they are also maintained in the system. `/etc/init.d/ucitrack info`

> > > Also I would drop the enable thing. If the entry doesn't need to be enabled, we just remove it.
> > 
> > 
> > But then the default value is no longer set as with the bootup, the configured value in the systctl still remains! Additional in other uci configuration files/section also have the disable uci feature. So If I do not want to apply this configuration I could set disable flag and do not have to remove the whole uci configuration section. I think this is a good useability.

[What about this. Can I do it like this. I think this is a good feature, especially as it is also done in other sections, e.g. 
 [network route uci option disabled](https://openwrt.org/docs/guide-user/network/ucicheatsheet#section_route), [network uci option disabled for interfaces](https://openwrt.org/docs/guide-user/network/ucicheatsheet#section_interface) or [firewall zone uci option enabled](https://openwrt.org/docs/guide-user/firewall/firewall_configuration#zones)
